USBDevice HAL specification

c1728p9 opened the pull request:
Initial USBDevice HAL specification changes

c1728p9 commented:
/morph build

0xc0170 commented on targets/TARGET_NXP/TARGET_LPC176X/usb/USBHAL_LPC17.cpp:
```diff
@@ -0,0 +1,660 @@
+    return true;
+}
+
+void USBPhyHw::endpoint_remove(usb_ep_t endpoint) {
```
`{` on the new line for functions (some other files have the same)

c1728p9 commented on usb/device/USBDevice/USBDevice_Types.h:
```diff
@@ -0,0 +1,86 @@
+    uint8_t  bRequest;
+    uint16_t wValue;
+    uint16_t wIndex;
+    uint16_t wLength;
```
I left these names unchanged as they matched the naming in the USB 2.0 specification. Do you think it would be clearer if I removed the size prefix? Let me know and I'll update this PR.

USB table with naming for reference:
![usb table](https://user-images.githubusercontent.com/2916772/36733706-0454aa10-1b97-11e8-9467-089f9c8d4010.PNG)

0xc0170 replied:
I was just curious as haven't seen these prefixes before. Fine as it is, thanks for the info

c1728p9 commented on usb/device/USBDevice/USBDevice.h:
```diff
@@ -0,0 +1,376 @@
+    }
+
+    /**
+     * Assert that the current thread of execution holds the lock
```
This is used to trap if a function that requires external synchronization is called without the lock being held. It is used in several places in USBDevice.cpp:
https://github.com/ARMmbed/mbed-os/pull/6215/files/d049ba9727313c4ef063429f6083187321305582#diff-1c41d1f789b2363f1020e89d0d62e422R800

c1728p9 commented:
/morph build

c1728p9 commented:
Changes made to this PR:
- ran astyle
- rebased to master
- added power change callback to phy

cmonr commented:
Failed to check out ARM license.
/morph build

adbridge commented:
Moving back to needs review as this has no approval from any of the reviewers!

c1728p9 commented:
Updated to include some fixes and cleanup

c1728p9 commented:
/morph build

c1728p9 commented:
@bulislaw or @0xc0170 can you add Filip and/or Maciej as reviewers as well?

c1728p9 commented:
Made the following changes:
- Move USBDevice enums and structs into USBDevice.h
- Added virtual power and state change callbacks to USBDevice
- Make reset callback optional (non abstract in USBDevice)
- Rebased to master

c1728p9 commented:
/morph build

maciejbocianski commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -0,0 +1,333 @@
+
+    # Control OUT stall
+    try:
+        request_type = 0x40                         # Host-to-device, Vendor request to Device
```
would be better to have less magic numbers, we can use `usb.util` stuff
```
# Host-to-device, Vendor request to Device
request_type = usb.util.build_request_type(
                                usb.util.CTRL_OUT, 
                                usb.util.CTRL_TYPE_VENDOR, 
                                usb.util.CTRL_RECIPIENT_DEVICE)
```

cmonr replied:
I would say the less magic numbers there are, the better.

c1728p9 replied:
done

c1728p9 commented:
/morph build
